p2p/gossip/protocol.go

```diff
@@ -0,0 +1,309 @@
+	"github.com/spacemeshos/go-spacemesh/p2p/message"
+	"github.com/spacemeshos/go-spacemesh/p2p/net"
+	"github.com/spacemeshos/go-spacemesh/p2p/node"
+	"sync"
+	"time"
+	"github.com/davecgh/go-spew/spew"
+)
+
+const PeerMessageQueueSize = 100
+
+type Protocol interface {
+	Broadcast(payload []byte) error
+	Start() error
+	Peer(pubkey string) (node.Node, net.Connection)
+	Shutdown()
+}
+
+type PeerSampler interface {
+	SelectPeers(count int) []node.Node
+}
+
+type ConnectionFactory interface {
+	GetConnection(address string, pk crypto.PublicKey) (net.Connection, error)
+}
+
+type Neighborhood struct {
+	log.Log
+
+	config config.SwarmConfig
+
+	peers        map[string]*peer
+	morePeersReq chan struct{}
+
+	oldMessageMu sync.RWMutex
+	oldMessageQ  map[string]struct{}
+
+	ps PeerSampler
+
+	cp ConnectionFactory
+
+	shutdown chan struct{}
+
+	peersMutex sync.RWMutex
+}
+
+func NewNeighborhood(config config.SwarmConfig, ps PeerSampler, cp ConnectionFactory, log2 log.Log) *Neighborhood {
+	return &Neighborhood{
+		Log:          log2,
+		config:       config,
+		morePeersReq: make(chan struct{}, config.RandomConnections),
+		peers:        make(map[string]*peer, config.RandomConnections),
+		oldMessageQ:  make(map[string]struct{}), // todo : remember to drain this
+		ps:           ps,
+		cp:           cp,
+	}
+}
+
+var _ Protocol = new(Neighborhood)
+
+type peer struct {
+	log.Log
+	node.Node
+	disc          chan error
+	connected     time.Time
+	conn          net.Connection
+	knownMessages map[string]struct{}
+	msgQ          chan []byte
+}
+
+func makePeer(node2 node.Node, c net.Connection, log log.Log) *peer {
+	return &peer{
+		log,
+		node2,
+		make(chan error,1),
+		time.Now(),
+		c,
+		make(map[string]struct{}),
+		make(chan []byte, PeerMessageQueueSize),
+	}
+}
+
+func (p *peer) send(message []byte) error {
+	if p.conn == nil || p.conn.Session() == nil {
+		return fmt.Errorf("the connection does not exist for this peer")
+	}
+	return p.conn.Send(message)
+}
+
+func (p *peer) addMessage(msg []byte) error {
+	// dont do anything if this peer know this msg
+	if _, ok := p.knownMessages[hex.EncodeToString(msg)]; ok {
+		return errors.New("already got this msg")
+	}
+
+	// check if connection and session are ok
+	c := p.conn
+	session := c.Session()
+	if c == nil || session == nil {
+		// todo: refresh Neighborhood or create session
+		return errors.New("no session")
+	}
+
+	data, err := message.PrepareMessage(session, msg)
+
+	if err != nil {
+		return err
+	}
+
+	select {
+	case p.msgQ <- data:
+
+	default:
+		return errors.New("Q was full")
+
+	}
+
+	return nil
+}
+
+func (p *peer) start(dischann chan struct{}) {
+	// check on new peers if they need something we have
+	//c := make(chan []string)
+	//t := time.NewTicker(time.Second * 5)
+	for {
+		select {
+		case m := <-p.msgQ:
+			err := p.send(m)
+			if err != nil {
+				// todo: handle errors
+				log.Error("Failed sending message to this peer %v", p.Node.PublicKey().String())
+				p.disc <- err
```

Comment:
If we failed sending the message, probably the connection isn't valid. we should inform `Neighborhood` so it will remove it from `peers` and won't try again. We should add a differentiation to when we failed because message format was bad.